File: 2/swift/main.swift
import Foundation

enum RawCommand: String, CaseIterable {
  case forward, down, up
}

struct Command {
  let rawCommand: RawCommand
  let amount: Int

  init?(_ rawValue: String) {
    for command in RawCommand.allCases {
      let prefix = "\(command.rawValue) "
      if rawValue.hasPrefix(prefix) {
        rawCommand = command
        let stringAmount = rawValue.dropFirst(prefix.count)
        if let intAmount = Int(stringAmount) {
          amount = intAmount
          return
        } else {
          return nil
        }
      }
    }

    return nil
  }

  var position: Int {
    switch rawCommand {
    case .forward:
      return amount
    case .down, .up:
      return 0
    }
  }

  var depth: Int {
    switch rawCommand {
    case .forward:
      return 0
    case .down:
      return amount
    case .up:
      return -amount
    }
  }
}

let contents = try String(contentsOfFile: "input.txt", encoding: .utf8)
let lines = contents.components(separatedBy: .newlines)
let commands = lines.compactMap(Command.init)

// MARK: - Part 1

print("# Part 1")

do {
  let (position, depth) = commands.reduce(into: (0, 0)) { result, command in
    result.0 += command.position
    result.1 += command.depth
  }

  print("Position: \(position)")
  print("Depth: \(depth)")
  print("Multiplied: \(position * depth)")
}

// MARK: - Part 2

print("# Part 2")

do {
  var position = 0
  var depth = 0
  var aim = 0

  for command in commands {
    position += command.position
    aim += command.depth
    depth += command.position * aim
  }

  print("Position: \(position)")
  print("Depth: \(depth)")
  print("Multiplied: \(position * depth)")
}

File: 2/cpp/main.cc
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
using namespace std;

enum RawCommand { forward, down, up };

struct Command {
  RawCommand raw_command;
  int amount;

  int position() {
    switch (raw_command) {
    case RawCommand::forward:
      return amount;
    case RawCommand::down:
      return 0;
    case RawCommand::up:
      return 0;
    }
  }

  int depth() {
    switch (raw_command) {
    case RawCommand::forward:
      return 0;
    case RawCommand::down:
      return amount;
    case RawCommand::up:
      return -amount;
    }
  }
};

int main() {
  ifstream is("input.txt");
  string str_command;

  vector<Command> commands;
  while(getline(is, str_command)) {
    string forward = "forward ";
    string down = "down ";
    string up = "up ";

    if (str_command.compare(0, forward.size(), forward) == 0) {
      Command command;
      command.raw_command = RawCommand::forward;
      int amount = stoi(str_command.substr(forward.size()));
      command.amount = amount;
      commands.push_back(command);
    } else if (str_command.compare(0, down.size(), down) == 0) {
      Command command;
      command.raw_command = RawCommand::down;
      int amount = stoi(str_command.substr(down.size()));
      command.amount = amount;
      commands.push_back(command);
    } else if (str_command.compare(0, up.size(), up) == 0) {
      Command command;
      command.raw_command = RawCommand::up;
      int amount = stoi(str_command.substr(up.size()));
      command.amount = amount;
      commands.push_back(command);
    }
  }

  {
    cout << "# Part 1" << endl;

    int position = 0;
    int depth = 0;

    for (auto i = commands.begin(); i != commands.end(); ++i) {
      position += i->position();
      depth += i->depth();
    }

    cout << "Position: " << position << endl;
    cout << "Depth: " << depth << endl;
    cout << "Multiplied: " << position * depth << endl;
  }

  {
    cout << "# Part 2" << endl;

    int position = 0;
    int depth = 0;
    int aim = 0;

    for (auto i = commands.begin(); i != commands.end(); ++i) {
      position += i->position();
      aim += i->depth();
      depth += i->position() * aim;
    }

    cout << "Position: " << position << endl;
    cout << "Depth: " << depth << endl;
    cout << "Multiplied: " << position * depth << endl;
  }
  return 0;
}
